test/topology_custom/test_topology_recovery_basic.py

```diff
@@ -0,0 +1,80 @@
+        delete_raft_data_and_upgrade_state, log_run_time, wait_until_upgrade_finishes as wait_until_schema_upgrade_finishes, \
+        wait_until_topology_upgrade_finishes, delete_raft_topology_state, check_system_topology_and_cdc_generations_v3_consistency
+
+
+@pytest.mark.asyncio
+@log_run_time
+async def test_topology_recovery_basic(request, manager: ManagerClient):
+    cfg = {'enable_user_defined_functions': False,
+           'experimental_features': ['consistent-topology-changes']}
+    servers = await manager.servers_add(3, config=cfg)
+    cql = manager.cql
+    assert(cql)
+
+    logging.info("Waiting until driver connects to every server")
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    logging.info(f"Restarting hosts {hosts} in recovery mode")
+    await asyncio.gather(*(enter_recovery_state(cql, h) for h in hosts))
+    # Restart sequentially, as it tests how nodes operating in legacy mode
+    # react to raft mode nodes and vice versa
+    for srv in servers:
+        await restart(manager, srv)
+    cql = await reconnect_driver(manager)
+
+    logging.info("Cluster restarted, waiting until driver reconnects to every server")
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+    logging.info(f"Driver reconnected, hosts: {hosts}")
+
+    logging.info(f"Deleting Raft data and upgrade state on {hosts}")
+    await asyncio.gather(*(delete_raft_topology_state(cql, h) for h in hosts))
+    await asyncio.gather(*(delete_raft_data_and_upgrade_state(cql, h) for h in hosts))
+
+    logging.info(f"Restarting hosts {hosts}")
+    for srv in servers:
+        await restart(manager, srv)
+    cql = await reconnect_driver(manager)
+
+    logging.info("Cluster restarted, waiting until driver reconnects to every server")
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    logging.info("Waiting until upgrade to raft schema finishes")
+    await asyncio.gather(*(wait_until_schema_upgrade_finishes(cql, h, time.time() + 60) for h in hosts))
+
+    logging.info("Checking the topology upgrade state on all nodes")
+    for host in hosts:
+        status = await manager.api.raft_topology_upgrade_status(host.address)
+        assert status == "not_upgraded"
+
+    logging.info("Waiting until all nodes see others as alive")
+    await asyncio.gather(*(manager.server_sees_others(srv.server_id, len(servers) - 1, time.time() + 60) for srv in servers))
+
+    logging.info("Triggering upgrade to raft topology")
+    await manager.api.upgrade_to_raft_topology(hosts[0].address)
+
+    logging.info("Waiting until upgrade finishes")
+    await asyncio.gather(*(wait_until_topology_upgrade_finishes(manager, h.address, time.time() + 60) for h in hosts))
+
+    logging.info("Checking consistency of data in system.topology and system.cdc_generations_v3")
+    await asyncio.gather(*(check_system_topology_and_cdc_generations_v3_consistency(manager, h) for h in hosts))
```

Comment:
Perhaps we could verify that the state is the same across nodes? (Which should be true without having to do read barriers from the test, as there is the global barrier done by coordinator already) Just checking if fields are `not None` doesn't seem to provide much value.